packages/taiko-client/driver/driver_test.go

```diff
@@ -344,6 +354,301 @@ func (s *DriverTestSuite) InitProposer() {
+		err    error
+	)
+	s.d.softblockServer, err = softblocks.New("*", nil, s.d.ChainSyncer().BlobSyncer(), s.RPCClient, true)
+	s.Nil(err)
+	go func() { s.NotNil(s.d.softblockServer.Start(port)) }()
+	defer func() { s.Nil(s.d.softblockServer.Shutdown(s.d.ctx)) }()
+
+	url, err := url.Parse(fmt.Sprintf("http://localhost:%v", port))
+	s.Nil(err)
+
+	headL1Origin, err := s.RPCClient.L2.HeadL1Origin(context.Background())
+	s.Nil(err)
+
+	for range headL1Origin.BlockID.Uint64() + epochs + 1 {
+		s.ProposeAndInsertEmptyBlocks(s.p, s.d.ChainSyncer().BlobSyncer())
+	}
+
+	l1Head, err := s.d.rpc.L1.HeaderByNumber(context.Background(), nil)
+	s.Nil(err)
+
+	l2Head, err := s.d.rpc.L2.HeaderByNumber(context.Background(), nil)
+	s.Nil(err)
+
+	for i := range epochs {
+		s.True(s.insertSoftBlock(url, l1Head, l2Head.Number.Uint64()+1, i, false, false).IsSuccess())
+	}
+	s.True(s.insertSoftBlock(url, l1Head, l2Head.Number.Uint64()+1, epochs, false, true).IsSuccess())
+	s.False(s.insertSoftBlock(url, l1Head, l2Head.Number.Uint64()+1, epochs+1, false, false).IsSuccess())
+
+	l2Head2, err := s.d.rpc.L2.HeaderByNumber(context.Background(), nil)
+	s.Nil(err)
+
+	// Remove soft blocks
+	res, err := resty.New().
+		R().
+		SetBody(&softblocks.RemoveSoftBlocksRequestBody{
+			NewLastBlockID: l2Head2.Number.Uint64() - 1,
+		}).
+		Delete(url.String() + "/softBlocks")
+	s.Nil(err)
+	s.True(res.IsSuccess())
+
+	l2Head3, err := s.d.rpc.L2.HeaderByNumber(context.Background(), nil)
+	s.Nil(err)
+	s.Equal(l2Head2.Number.Uint64()-1, l2Head3.Number.Uint64())
+}
+
 func TestDriverTestSuite(t *testing.T) {
 	suite.Run(t, new(DriverTestSuite))
 }
+
+// insertSoftBlock inserts a soft block with the given parameters.
+func (s *DriverTestSuite) insertSoftBlock(
+	url *url.URL,
+	anchoredL1Block *types.Header,
+	l2BlockID uint64,
+	batchID uint64,
+	endOfBlock bool,
+	endOfPreconf bool,
+) *resty.Response {
+	preconferPrivKey, err := crypto.ToECDSA(common.FromHex(os.Getenv("L1_PROPOSER_PRIVATE_KEY")))
+	s.Nil(err)
+
+	preconferAddress := crypto.PubkeyToAddress(preconferPrivKey.PublicKey)
+
+	nonce, err := s.RPCClient.L2.NonceAt(context.Background(), s.TestAddr, nil)
+	s.Nil(err)
+
+	tx := types.NewTransaction(
+		nonce,
+		common.BytesToAddress(testutils.RandomBytes(32)),
+		common.Big0,
+		100_000,
+		new(big.Int).SetUint64(uint64(10*params.GWei)),
+		[]byte{},
+	)
+	signedTx, err := types.SignTx(tx, types.LatestSignerForChainID(s.RPCClient.L2.ChainID), s.TestAddrPrivKey)
+	s.Nil(err)
+
+	// If the transaction is underpriced, we just ingore it.
+	err = s.RPCClient.L2.SendTransaction(context.Background(), signedTx)
+	if err != nil {
+		s.Equal("replacement transaction underpriced", err.Error())
+	}
+
+	b, err := encodeAndCompressTxList([]*types.Transaction{signedTx})
+	s.Nil(err)
+
+	var marker softblocks.TransactionBatchMarker
+	if endOfBlock {
+		marker = softblocks.BatchMarkerEOB
+	} else if endOfPreconf {
+		marker = softblocks.BatchMarkerEOP
```

Comment:
What is the difference between the EOB and EOP?

Reply:
EoB means this soft block is ended, can't append new batches, EoP means this current preconfer is done, can't append new batches and new soft blocks.